### PR TITLE
W-12049491: Added new library to support Draft 6, 7, 2019-09 and 2020-12

### PR DESCRIPTION
**Pull request by GuidoGianfrate**
*Hi, thank you for your work. We understand that you want to merge your changes and move on from this issue, but we also want to maintain the safety, readability, and testability of our code. Because of this, we kindly ask that you confirm that you have fulfilled the prerequisites for each category of activity before merging your changes.*

### If the Pull Request has a dependency update:
- [ ] I have read the Release Notes for the new version (and intermediate versions if the jump would include more than one). Don't blindly trust that the dependency will honor SEMVER, always read carefully the release notes
- [ ] Java 8+ support is maintained
- [ ] The new version has no vulnerabilities
- [ ] A team member reviewed the Pull Request.
- [ ] No backward compatibility is broken.
- [ ] If it is a "provided" dependency, it has been checked that if it is suggested, the version number is the same.

### For ALL the Releases:
- [ ] There is a task in GUS for this change.
- [ ] The corresponding Release Notes have been written and shared with the documentation team.
- [ ] The new code has tests and they have before and after methods to be sure that you are working with a clean environment and that you are leaving the environment clean after they finish.
- [ ] If you are using reflection, create a test to call the methods you are invoking, this way, if there is a change in the API, the test will be able to detect it.
- [ ] If the project has a parent pom and modules, the release pipeline only uploads the modules, so, please, make sure that the parent was deployed in the repositories before considering this release finished.
- [ ] Notify in the Slack Channel that the release is complete, so the Docs team can merge the Documentation and Release Notes PR. The docs team always checks in Anypoint Exchange that the connector is released before merging the release notes.
- [ ] Check the pending Pull Requests in the project coming from Renovate to see if it is possible to update some dependencies. (For those that you cannot merge, follow the same procedure as for the [periodic renovate review process](https://confluence.internal.salesforce.com/display/CONNECTORS/MuleSoft+Connectivity+Infrastructure+Documentation#MuleSoftConnectivityInfrastructureDocumentation-RenovatePullRequests)).
 
**NOTE**: *The release process ends when you merge the pull request that updates the support branch to the new snapshot, please don't forget to do this.*

### Patch Version:
- [ ] The Pull Request has been peer-reviewed.
- [ ] No backward compatibility is broken.
- [ ] Documentation: (Required) Release Notes PR with compatibility table. Share it with the Docs team.

### Minor Version:
- [ ] Have a document with the specification of the release.
- [ ] Create a slack channel (eg: rl-conninfra-sftp-1.1.1-new-chiper) to coordinate the release of this version with the documentation team and the architects of our team (at least one week before the release). Share the spec doc with the docs team.
- [ ] The documentation fork has been done and has been updated with the changes related to the new functionalities.
- [ ] The pull request has been reviewed by a peer and the team leader.
- [ ] Public Documentation: Minor releases nearly always require a revision of the documentation because the release usually includes new features as well as bug fixes. Even if there are no user-facing changes, documentation has to be versioned for every minor release.
    - [ ] (Required) Release Notes PR with compatibility table. Share it with the Docs team.
    - [ ] (Likely required) Reference guide. Run the script and share the file with the Docs team.
    - [ ] (Assess) Often a minor release impacts UI and examples, so the documentation (user guides, examples, troubleshooting guides) should be reviewed and updated accordingly. Notify the docs team if any updates are needed.

### Major Version:
- [ ] Have a document with the specification of the release.
- [ ] Create a slack channel (eg: rl-conninfra-sftp-2.0.0) to coordinate the release of this version with the documentation team and the architects of our team (at least one month before the release). Share the spec doc with the docs team.
- [ ] The documentation fork has been made and updated with the changes related to the new functionalities and possible compatibility problems with the previous version.
- [ ] The systems properties have been removed, detailing in the documentation the changes in the default behavior of the product.
- [ ] The TODO comments of the code have been reviewed
- [ ] A colleague, the team leader, and the team architect have reviewed the Pull Request.
- [ ] Public Documentation: Major releases break backward compatibility and should never be released without documentation, as this breaks the user experience.
    - [ ] (Required) Upgrade and Migration Guide: Every major release must have a detailed upgrade guide (template) with information about what was changed and what steps the user has to take to ensure the connector works as expected. Share the details with the Docs team so they can update the guide.
    - [ ] (Required) Release Notes PR with compatibility table. Share it with the docs team.
    - [ ] (Required) Reference guide. Run the script and share the file with the Docs team.
    - [ ] User guide updates, new screenshots or examples (if applicable), share the updates with the docs team.

### CODE REVIEW

**Review by acurci-at-mulesoft — DISMISSED** [review]

File: src/main/java/org/mule/module/json/internal/ValidatorCommonUtils.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.net.URI;
+import java.net.URL;
+
+/**
+ * Common methods that are useful for the other classes
+ */
+public class ValidatorCommonUtils {
+
+  private ValidatorCommonUtils() {}
```
Comment (elcoya) [inline src/main/java/org/mule/module/json/internal/ValidatorCommonUtils.java:17]: Shouldn't this class be done static instead of making the constructor private?

File: src/main/java/org/mule/module/json/internal/ValidationLibraries.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+ * LICENSE.txt file.
+ */
+package org.mule.module.json.internal;
+
+/**
+ * Libraries that can be used to make validations
+ */
+public enum ValidationLibraries {
+  JAVA_JSON_TOOLS, NETWORKNT
+}
```
Comment (elcoya) [inline src/main/java/org/mule/module/json/internal/ValidationLibraries.java:14]: Consider adding a constructor with the supported version for each behavior.

File: src/main/java/org/mule/module/json/internal/JsonSchemaValidator.java [hunk 3]
```diff
@@ -7,356 +7,102 @@
...
-
-  /**
-   * Returns a new {@link Builder}
-   *
-   * @return a {@link Builder}
-   */
-  public static Builder builder() {
-    return new Builder();
+  public void validate(InputStream inputStream) {
+    // do nothing
```
Comment (elcoya) [inline src/main/java/org/mule/module/json/internal/JsonSchemaValidator.java:72]: Is this really the default behavior or should this be and abstract method?

**Review by elcoya — COMMENTED** [review]
@GuidoGianfrate please review my comments

File: src/main/java/org/mule/module/json/internal/JsonSchemaValidator.java (hunk 3 above)
Comment (GuidoGianfrate) [reply src/main/java/org/mule/module/json/internal/JsonSchemaValidator.java:72]: Should be an abstract method 👍

File: src/main/java/org/mule/module/json/internal/JsonSchemaValidatorJavaJsonToolsWrapper.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+  @Override
+  public void validate(InputStream inputStream) {
+    JsonNode jsonNode = super.asJsonNode(inputStream);
+    ProcessingReport report;
+    try {
+      report = jsonSchema.validate(jsonNode);
+
+    } catch (ProcessingException processingException) {
+      //TODO We must create a new error: INVALID_REFERENCE, to inform the user that the external references declared in the Schema cannot be accessed(W-12301483)
```
Comment (elcoya) [inline src/main/java/org/mule/module/json/internal/JsonSchemaValidatorJavaJsonToolsWrapper.java:59]: Is there any reason not to complete the TODO before merging? @GuidoGianfrate

Comment (GuidoGianfrate) [reply src/main/java/org/mule/module/json/internal/JsonSchemaValidatorJavaJsonToolsWrapper.java:59]: Yes, because this TODO is for the next MAJOR

File: src/main/java/org/mule/module/json/internal/JsonSchemaValidatorNetworkntWrapper.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+  @Override
+  public void validate(InputStream inputStream) {
+    JsonNode jsonNode = super.asJsonNode(inputStream);
+    Set<ValidationMessage> responseValidate;
+    try {
+      responseValidate = jsonSchema.validate(jsonNode);
+    } catch (Exception e) {
+
+      if (e.getMessage().contains("Reference") && e.getMessage().contains("cannot be resolved")) {
```
[the remark this answers is not in the document]
Comment (elcoya) [reply src/main/java/org/mule/module/json/internal/JsonSchemaValidatorNetworkntWrapper.java:51]: I checked the regex `Reference.+cannot be resolved` on https://regex101.com/ and it would seem to match correctly

File: src/main/java/org/mule/module/json/internal/JsonSchemaValidatorJavaJsonToolsWrapper.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+      } else {
+        throw new MuleRuntimeException(createStaticMessage(
+                                                           "Exception was found while trying to validate against json schema. Content was: "
+                                                               + jsonNode.toString()),
+                                       processingException);
+      }
+    }
+
+    if (!report.isSuccess()) {
+      throw new SchemaValidationException("Json content is not compliant with schema: \n" + report,
```
Comment (elcoya) [inline src/main/java/org/mule/module/json/internal/JsonSchemaValidatorJavaJsonToolsWrapper.java:71]: Use a constant definition for the error message please.
```
public static final String VALIDATION_FAILED_MESSAGE = "Json content is not compliant with schema.\n"

File: src/main/java/org/mule/module/json/internal/JsonSchemaValidatorNetworkntWrapper.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      JsonSchemaFactory schemaFactory = JsonSchemaFactory.getInstance(SpecVersionDetector.detect(jsonNode));
+
+      if (schemaLocation == null) {
+        return schemaFactory.getSchema(jsonNode, getUriRedirectConfig(schemaRedirects));
+      } else {
+        return schemaFactory.getSchema(new URI(resolveLocationIfNecessary(schemaLocation)),
+                                       getUriRedirectConfig(schemaRedirects));
+      }
+    } catch (URISyntaxException e) {
+      throw new ModuleException(format("Could not load JSON schema [%s]. %s", schemaLocation, e.getMessage()),
```
Comment (elcoya) [inline src/main/java/org/mule/module/json/internal/JsonSchemaValidatorNetworkntWrapper.java:87]: Please review the case: This exception message does not seem related to the one that can be thrown when using the JavaJsonTool lib. Can we make the cases more similar? They don't look like catching the same problems.

File: src/main/java/org/mule/module/json/internal/LibraryLink.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+package org.mule.module.json.internal;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+/**
+ * Defines the default behaviour to implement chain of responsibility design pattern
+ */
+public abstract class LibraryLink {
+
+  LibraryLink nextLink;
```
Comment (elcoya) [inline src/main/java/org/mule/module/json/internal/LibraryLink.java:16]: Just in case, to avoid having the subclasses touching this reference, I would suggest to make this private.

**Review by elcoya — CHANGES_REQUESTED** [review]
I left some comments to review

File: src/main/java/org/mule/module/json/internal/JsonSchemaValidatorNetworkntWrapper.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  @Override
+  public void validate(InputStream inputStream) {
+    JsonNode jsonNode = super.asJsonNode(inputStream);
+    Set<ValidationMessage> responseValidate;
+    try {
+      responseValidate = jsonSchema.validate(jsonNode);
+    } catch (Exception e) {
+
+      if (e.getMessage().contains("Reference") && e.getMessage().contains("cannot be resolved")) {
+        //TODO We must create a new error: INVALID_REFERENCE, to inform the user that the external references declared in the Schema cannot be accessed(W-12301483)
```
[the remark this answers is not in the document]
Comment (GuidoGianfrate) [reply src/main/java/org/mule/module/json/internal/JsonSchemaValidatorNetworkntWrapper.java:52]: This TODO already exist in main branch, we agree with the team that throw a new type of error is for the next Major.
This is a Minor.

File: src/main/java/org/mule/module/json/internal/JsonSchemaValidatorJavaJsonToolsWrapper.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    } catch (ProcessingException processingException) {
+      //TODO We must create a new error: INVALID_REFERENCE, to inform the user that the external references declared in the Schema cannot be accessed(W-12301483)
+      if (processingException.getClass().getName().contains("ProcessingException")) {
+        throw new MuleRuntimeException(createStaticMessage("Invalid Schema References"), processingException);
+      } else {
+        throw new MuleRuntimeException(createStaticMessage(
+                                                           "Exception was found while trying to validate against json schema. Content was: "
+                                                               + jsonNode.toString()),
+                                       processingException);
+      }
```
[the remark this answers is not in the document]
Comment (elcoya) [reply src/main/java/org/mule/module/json/internal/JsonSchemaValidatorJavaJsonToolsWrapper.java:67]: As of our later discussion, the recommended approach goes like follows
Declare a constant Pattern in the class
```
private static final Pattern INVALID_REFERENCE_MESSAGE_PATTERN = Pattern.compile(".*invalid reference.*");
```

and then, just use the matcher
```
      Matcher messageMatcher = INVALID_REFERENCE_MESSAGE_PATTERN.matcher(processingException.getMessage());
      if (messageMatcher.matches()) {
        throw new MuleRuntimeException(createStaticMessage("Invalid Schema References"), processingException);
      } else {
        throw new MuleRuntimeException(createStaticMessage(
                "Exception was found while trying to validate against json schema. Content was: "
                        + jsonNode.toString()),
                processingException);
      }
```

**Review by elcoya — DISMISSED** [review]
LGTM

**Review by elcoya — APPROVED** [review]
still good
